ateam_vision_filter/src/filters/kalman_filter.hpp

```diff
@@ -0,0 +1,59 @@
+// Copyright 2021 A Team
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#ifndef FILTERS__KALMAN_FILTER_HPP_
+#define FILTERS__KALMAN_FILTER_HPP_
+
+#include <Eigen/Dense>
+
+// Standard kalman filter
+
+class KalmanFilter
+{
+public:
+  void set_initial_x_hat(const Eigen::VectorXd & x_hat);
+  void set_initial_p(const Eigen::MatrixXd & P);
+  void set_F(const Eigen::MatrixXd & F);
+  void set_B(const Eigen::MatrixXd & B);
+  void set_H(const Eigen::MatrixXd & H);
+  void set_Q(const Eigen::MatrixXd & Q);
+  void set_R(const Eigen::MatrixXd & R);
+
+  void predict(const Eigen::VectorXd & u);
+  void update(const Eigen::VectorXd & z);
+
+  Eigen::VectorXd get_x_hat() const;
+  Eigen::MatrixXd get_P_hat() const;
+  Eigen::VectorXd get_y() const;
+  Eigen::MatrixXd get_estimated_gaussian_variance() const;
+  Eigen::VectorXd get_potential_measurement_error(const Eigen::VectorXd & z);
+
+private:
+  Eigen::MatrixXd F;
+  Eigen::MatrixXd B;
+  Eigen::MatrixXd H;
+  Eigen::MatrixXd Q;
+  Eigen::MatrixXd R;
+
+  Eigen::VectorXd x_hat;
+  Eigen::MatrixXd P;
```

Comment:
I like that perspective as well.

I was hitting some annoyances with how to set this up programmatically without copy-pasting tons of code or over engineering it. Essentially, each camera has a list of MHT for each object (32 robot MHT's, and 1 ball MHT). In each MHT, each track is a IMM. Each model in the IMM is represented by a different KF. (This isn't super clear rereading it, I can draw a picture if it doesn't make sense).

The only difference between a robot MHT and a ball MHT is the base state space model, the list of model types being used, the input generator, and the probability generator.

The proper way to do it all would be to do dependency injection, but that feels a little too over complicate since there will only ever be two options. Having if statements everywhere isn't the most clear either. I half-assed the dependency injection by combining the input and probability generators, making everything work on Eigen::VectorXd instead of specific sizes, and passing in a specific kalman filter instantiation that will be cloned whenever we need the kalman filter.

Thoughts? It's setup in a way that it will be easy to transition in almost any direction.